internal/cri/server/container_create.go

```diff
@@ -362,6 +373,99 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
 	return &runtime.CreateContainerResponse{ContainerId: id}, nil
 }
 
+func (c *criService) mutateMounts(
+	ctx context.Context,
+	platform imagespec.Platform,
+	extraMounts []*runtime.Mount,
+) error {
+	for _, m := range extraMounts {
+		err := c.mutateOCIBindMount(ctx, platform, m)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *criService) mutateOCIBindMount(
+	ctx context.Context,
+	platform imagespec.Platform,
+	extraMount *runtime.Mount,
+) (retErr error) {
+	imageSpec := extraMount.GetImage()
+	if imageSpec == nil {
+		return nil
+	}
+
+	ref := imageSpec.GetImage()
+	if ref == "" {
+		return fmt.Errorf("image not specified")
+	}
+
+	image, err := c.LocalResolve(ref)
+	if err != nil {
+		return fmt.Errorf("failed to resolve image %q: %w", ref, err)
+	}
+	target := c.getImageVolumeDir(strings.TrimPrefix(image.ID, "sha256:"))
+
+	snapshotter := defaults.DefaultSnapshotter
+
+	ctx, done, err := c.client.WithLease(ctx,
+		leases.WithID(target),
+		leases.WithExpiration(24*time.Hour),
+		leases.WithLabel("containerd.io/gc.ref.snapshot."+snapshotter, target),
+	)
+	if err != nil && !errdefs.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create lease %q: %w", target, err)
+	}
+	defer func() {
+		if retErr != nil && done != nil {
+			done(ctx)
+		}
+	}()
+
+	img, err := c.client.ImageService().Get(ctx, ref)
+	if err != nil {
+		return fmt.Errorf("failed to get image %q: %w", ref, err)
+	}
+
+	i := containerd.NewImageWithPlatform(c.client, img, platforms.Only(platform))
+	if err := i.Unpack(ctx, snapshotter); err != nil {
+		return fmt.Errorf("failed to unpacking image: %w", err)
+	}
+
+	diffIDs, err := i.RootFS(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to getting diff IDs for image %q: %w", ref, err)
+	}
+	chainID := identity.ChainID(diffIDs).String()
+
+	s := c.client.SnapshotService(snapshotter)
+	mounts, err := s.View(ctx, target, chainID)
+	if err != nil {
+		if errdefs.IsAlreadyExists(err) {
+			mounts, err = s.Mounts(ctx, target)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to get mounts for image %q: %w", ref, err)
+		}
+	}
+
+	err = os.MkdirAll(target, 0755)
+	if err != nil {
+		return fmt.Errorf("failed to creating directory %q: %w", target, err)
+	}
+
+	if err := mount.All(mounts, target); err != nil {
+		_ = s.Remove(ctx, target)
+		return fmt.Errorf("failed to mounting %q: %w", target, err)
+	}
+
+	extraMount.HostPath = target
+	extraMount.Readonly = true
```

Comment:
https://github.com/cri-o/cri-o/pull/8317/files#diff-1e0877112035840d3e2f073c693665f8daffa241a440e54a866ad7382b510ea2R1077-R1082

Refer to the cri-o to change the HostPath directly.

Reply:
see other comment.. test host path is unset before assignment.. and validate readonly